Integrate Interchain Accounts into Gravity Bridge

ChristianBorst opened the pull request:
A slight cleanup of @gadost 's amazing work to integrate ica into Gravity-Bridge.

@jkilpatr I added some comments in here to help with understanding during the review. Hopefully you won't necessarily need to read the whole ICA spec to understand what the functionality and flow is, but I am always available for questions.

gadost committed:
Add ICA, upgrade, and integration test

* ICA Controller, Host module
* icaauth module
* ICA upgrade preparation
* ICA integration test

ChristianBorst commented on .github/workflows/integration-tests.yml:
```diff
@@ -2,9 +2,9 @@ name: Integration tests
 
 on:
   push:
-    branches: [master, main]
```
TODO: Remove this

ChristianBorst commented on module/app/app.go:
```diff
@@ -210,28 +231,34 @@ type Gravity struct {
+	crisisKeeper        *crisiskeeper.Keeper
+	upgradeKeeper       *upgradekeeper.Keeper
+	paramsKeeper        *paramskeeper.Keeper
+	ibcKeeper           *ibckeeper.Keeper
```
Added all 3 ica keepers here

ChristianBorst commented on module/app/app.go:
```diff
@@ -984,4 +1094,23 @@ func (app *Gravity) registerStoreLoaders() {
+				Added: []string{icacontrollertypes.StoreKey,
+					icahosttypes.StoreKey,
+					icaauthtypes.StoreKey},
+				// Check upgrade docs to see which type of store loader is necessary for deletes/renames
```
This upgrade will add 3 new stores to the chain

ChristianBorst commented on module/app/upgrades/icav3/handler.go:
```diff
@@ -0,0 +1,65 @@
+				sdk.MsgTypeURL(&distrtypes.MsgWithdrawValidatorCommission{}),
+				sdk.MsgTypeURL(&distrtypes.MsgFundCommunityPool{}),
+				sdk.MsgTypeURL(&govtypes.MsgVote{}),
+			},
```
These are the accepted messages the ICA host module can receive. By default the ICA controller and icaauth modules will submit any message types they have registered in their codecs, but the ICA host module will reject anything not in this specific AllowMessages list.

Visually:
[ **Chain A** ICAAuth SubmitTx ✅ -> ICA Controller ✅ ]-> Relayer -> [ **Chain B** ICA Host ❌ Rejection Here ]

ChristianBorst commented on module/third_party/proto/ibc/applications/interchain_accounts/controller/v1/controller.proto:
```diff
@@ -0,0 +1,14 @@
+syntax = "proto3";
```
The IBC protos (including ics23) all come from what ibc-go v3.2.1 is using

ChristianBorst commented on module/x/icaauth/ibc_module.go:
```diff
@@ -0,0 +1,144 @@
+	counterparty channeltypes.Counterparty,
+	version string,
+) error {
+	// Claim channel capability passed back by IBC module
```
Here is the only handshake related activity the icaauth module is supposed to perform, the icaauth module holds the capability for the Gravity side `PortId: icacontroller-gravity123456, ChannelId: channel-12` end of the connection. This channel will connect to the foreign chain with an end like `PortId: icahost, ChannelId: channel-34`

ChristianBorst commented on module/x/icaauth/keeper/msg_server.go:
```diff
@@ -0,0 +1,59 @@
+}
+
+// SubmitTx implements the Msg/SubmitTx interface
+func (k msgServer) SubmitTx(goCtx context.Context, msg *types.MsgSubmitTx) (*types.MsgSubmitTxResponse, error) {
```
MsgSubmitTx will contain an array of Msgs which should be submitted to the foreign chain's ICA host module and executed as the interchain account.

ChristianBorst commented on module/x/icaauth/keeper/account.go:
```diff
@@ -0,0 +1,40 @@
+
+// RegisterInterchainAccount invokes the InitInterchainAccount entrypoint.
+// InitInterchainAccount binds a new controller port and initiates a new ICS-27 channel handshake
+func (k Keeper) RegisterInterchainAccount(ctx sdk.Context, owner sdk.AccAddress, connectionID string, version string) error {
```
In the event of a channel failure (the channel is ordered, and might close at any time) this should not be initiated again instead a MsgChanOpenInit needs to be submitted to the controller chain.

ChristianBorst commented on orchestrator/gravity_utils/src/types/batches.rs:
```diff
@@ -17,7 +17,7 @@ pub struct BatchTransaction {
     /// This transactions Cosmos pool id
     pub id: u64,
     /// The senders Cosmos address
```
TODO: This needs to be reverted in favor of updating deep_space to include > 32 length address bytes

jkilpatr replied:
looks like a few things here need deep space upstreaming.

ChristianBorst commented on orchestrator/test_runner/src/airdrop_proposal.rs:
```diff
@@ -31,7 +31,9 @@ pub async fn airdrop_proposal_test(contact: &Contact, keys: Vec<ValidatorKeys>)
     // we start by disabling inflation, this allows us to check that the correct
     // amount of tokens was withdrawn from the community pool, if we don't do this inflation
     // in the community pool may be larger than our actual airdrop amount spent.
-    disable_inflation(contact, &keys).await;
```
Here I refactored the vote_yes_on_proposals() function so that we can submit proposals to the ibc chain, which doesn't have the necessary &[ValidatorKeys] type, and instead has &[CosmosPrivateKey]

ChristianBorst commented on orchestrator/test_runner/src/bootstrapping.rs:
```diff
@@ -301,7 +303,7 @@ pub fn create_ibc_channel(hermes_base: &mut Command) {
         "create",
         "channel",
         &get_gravity_chain_id(),
```
Hermes was updated to v0.15 for ICA support, some of the interfaces changed and new features need to be set up to enable ICA.

ChristianBorst commented on orchestrator/test_runner/src/ica.rs:
```diff
@@ -0,0 +1,816 @@
+    gravity_address: EthAddress,
+    web30: &Web3,
+    grpc_client: GravityQueryClient<Channel>,
+) {
```
This test needs just a little bit of work to clarify what is happening when viewing the logs, but it is a great test of base ICA functionality.

To summarize the test, it initializes the gravity and ibc chain's ICA Host AllowMessages parameter, creates an interchain account on each of gravity and the ibc chain, sends those interchain accounts some "stake" funds with a basic bank/MsgSend, and uses ICA to delegate to validator 0 from gravity and from ibc-chain, respectively.

gadost replied:
Also send to eth from ibc chain ( if MsgSendToEth added to codec on ibc chain  )

ChristianBorst commented on orchestrator/test_runner/src/ica.rs:
```diff
@@ -0,0 +1,816 @@
+
+    // Add allow messages to both chains
+    add_ica_host_allow_messages(contact, &val_priv_keys).await;
+    add_ica_host_allow_messages(cpc_contact, &ibc_keys).await;
```
The althea-net/gaia-ica repo differs from gadost's gaia repo, so I added a governance vote to enable all messages on the ibc-chain's ICA Host module.

ChristianBorst commented on tests/dockerfile/Dockerfile:
```diff
@@ -6,12 +6,13 @@ RUN dnf install -y git make gcc gcc-c++ which iproute iputils procps-ng vim-mini
-RUN cd /tmp/ && tar -xvf hermes-v0.13.0-x86_64-unknown-linux-gnu.tar.gz && mv hermes /usr/bin/
-RUN mkdir /ibc-relayer-logs && touch /ibc-relayer-logs/hermes-logs && touch /ibc-relayer-logs/channel-creation
+# Download modified gaia as a IBC test chain, IMO should be replaced with git clone or something
+ADD https://github.com/althea-net/gaia-ica/releases/download/v7.0.3/gaiad-v7.0.3-linux-amd64 /usr/bin/gaiad
```
This is the link to the althea-net repo I created to mimic the work gadost did. Gadost had a separate copy of the icaauth module that this branch of Gravity now contains, I simplified the changes by simply importing this branch of Gravity into althea-net/gaia-ica. Once this branch gets merged I will need to clean up that repo to point to main, or some particular version of Gravity.

jkilpatr commented on module/app/app.go:
```diff
@@ -539,7 +634,11 @@ func NewGravityApp(
+		AddRoute(icahosttypes.SubModuleName, icaHostIBCModule).
+		AddRoute(ibctransfertypes.ModuleName, ibcTransferIBCModule).
+		AddRoute(icaauthtypes.ModuleName, icaControllerIBCModule)
+
```
maybe put a comment in for this?

jkilpatr commented on orchestrator/gravity_proto/src/lib.rs:
```diff
@@ -9,3 +9,6 @@ pub mod gravity {
     include!("prost/gravity.v1.rs");
     include!("ethereum_claim.rs");
 }
+pub mod icaauth {
```
shouldn't this be upstream in deep space?

jkilpatr reviewed (commented):
This is really impressive work! The ICA test is impressive. Dropping a few comments.

ChristianBorst committed:
Remove unused SignType proto

ChristianBorst committed:
Bump althea-gaia to v7.0.4

ChristianBorst committed:
Fix ICA - dragonberry issue

With the release of sdk v0.45.9 and the dragonberry vulnerability fix,
the ibc-go repo also needs to be updated to fix an issue in interchain
account creation, to do so we switch to the althea-net/ibc-go fork

ChristianBorst committed:
Add ICA simple attacks test, problematic node halting test

ChristianBorst committed:
Update deep_space to 2.15

ChristianBorst committed:
TEST: Try to break CI with ica node wrecker test
